data/basics/teal.go

```diff
@@ -0,0 +1,298 @@
+			}
+		case SetUintAction:
+		case DeleteAction:
+		default:
+			return fmt.Errorf("unknown delta action")
+		}
+	}
+	return nil
+}
+
+// EvalDelta stores StateDeltas for an application's global key/value store, as
+// well as StateDeltas for some number of accounts holding local state for that
+// application
+type EvalDelta struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	GlobalDelta StateDelta `codec:"gd"`
+
+	// When decoding EvalDeltas, the integer key represents an offset into
+	// [txn.Sender, txn.Accounts[0], txn.Accounts[1], ...]
+	LocalDeltas map[uint64]StateDelta `codec:"ld,allocbound=config.MaxEvalDeltaAccounts"`
+}
+
+// Equal compares two EvalDeltas and returns whether or not they are
+// equivalent. It does not care about nilness equality of LocalDeltas,
+// because the msgpack codec will encode/decode an empty map as nil, and we want
+// an empty generated EvalDelta to equal an empty one we decode off the wire.
+func (ed EvalDelta) Equal(o EvalDelta) bool {
+	// GlobalDeltas must be equal
+	if !ed.GlobalDelta.Equal(o.GlobalDelta) {
+		return false
+	}
+
+	// LocalDeltas length should be the same
+	if len(ed.LocalDeltas) != len(o.LocalDeltas) {
+		return false
+	}
+
+	// All keys and local StateDeltas should be the same
+	for k, v := range ed.LocalDeltas {
+		if !o.LocalDeltas[k].Equal(v) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// StateSchema sets maximums on the number of each type that may be stored
+type StateSchema struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	NumUint      uint64 `codec:"nui"`
+	NumByteSlice uint64 `codec:"nbs"`
+}
+
+// AddSchema adds two StateSchemas together
+func (sm StateSchema) AddSchema(osm StateSchema) (out StateSchema) {
+	out.NumUint = AddSaturate(sm.NumUint, osm.NumUint)
+	out.NumByteSlice = AddSaturate(sm.NumByteSlice, osm.NumByteSlice)
+	return
+}
+
+// SubSchema subtracts one StateSchema from another
+func (sm StateSchema) SubSchema(osm StateSchema) (out StateSchema) {
+	out.NumUint = SubSaturate(sm.NumUint, osm.NumUint)
+	out.NumByteSlice = SubSaturate(sm.NumByteSlice, osm.NumByteSlice)
+	return
+}
+
+// NumEntries counts the total number of values that may be stored for particular schema
+func (sm StateSchema) NumEntries() (tot uint64) {
+	tot = AddSaturate(tot, sm.NumUint)
+	tot = AddSaturate(tot, sm.NumByteSlice)
+	return tot
+}
+
+// MinBalance computes the MinBalance requirements for a StateSchema based on
+// the consensus parameters
+func (sm StateSchema) MinBalance(proto *config.ConsensusParams) (res MicroAlgos) {
+	// Flat cost for each key/value pair
+	flatCost := MulSaturate(proto.SchemaMinBalancePerEntry, sm.NumEntries())
+
+	// Cost for uints
+	uintCost := MulSaturate(proto.SchemaUintMinBalance, sm.NumUint)
+
+	// Cost for byte slices
+	bytesCost := MulSaturate(proto.SchemaBytesMinBalance, sm.NumByteSlice)
+
+	// Sum the separate costs
+	var min uint64
+	min = AddSaturate(min, flatCost)
```

Comment:
nit:
min := flatCost.Raw

Reply:
Personally I find the way it is more clear, since each subsequent that adds to min is exactly the same. I can change it if you feel strongly